numba/tests/test_parfors.py

```diff
@@ -4307,5 +4438,70 @@ def strip_instrs(asm):
                         self.assertEqual(b[op[-2]:op[-1]], 'u')
 
 
+@skip_parfors_unsupported
+class TestParforReductionSetNumThreads(TestCase):
+    """Test exceution correctness on reductions with set_num_threads.
+    """
+    def test_add(self):
+        N = config.NUMBA_NUM_THREADS
+        M = 2 * N
+        mask = N - 1
+
+        @njit(parallel=True)
+        def udt(nthreads):
+            acc = 0
+            set_num_threads(nthreads)
+            for i in prange(M):
+                local_mask = 1 + i % mask
+                set_num_threads(local_mask)
+                gnt = get_num_threads()
+                acc += gnt
+            return acc
+
+        expect = udt.py_func(mask)
+        got = udt(mask)
+        self.assertPreciseEqual(expect, got)
+
+    def test_mul(self):
+        N = config.NUMBA_NUM_THREADS
+        M = 2 * N
+        mask = N - 1
+
+        @njit(parallel=True)
+        def udt(nthreads):
+            acc = 1
+            set_num_threads(nthreads)
+            for i in prange(M):
+                local_mask = 1 + i % mask
+                set_num_threads(local_mask)
+                gnt = get_num_threads()
+                acc *= gnt
```

Comment:
For a machine with e.g. 64 cores, this will overflow?

Reply:
Maybe "acc = acc % (2**64)" at the end of each iteration?

Reply:
Perhaps bound it like `N = min(4, config.NUMBA_NUM_THREADS)`?

Reply:
Done.